### PR TITLE
Ensure at least one vpub is zero

### PR DESCRIPTION
**Pull request by defuse**
Supersedes #895. This will conflict with #889 (and maybe others) but hopefully keeping the file from this PR will work with the other PR.

**WARNING: I force pushed.**

### CODE REVIEW

File: src/gtest/test_checktransaction.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+TEST(checktransaction_tests, check_vpub_not_both_nonzero) {
+    CMutableTransaction tx;
+    tx.nVersion = 2;
+
+    {
+        // Ensure that values within the pour are well-formed.
+        CMutableTransaction newTx(tx);
+        CValidationState state;
+        state.SetPerformPourVerification(false); // don't verify the snark
```
Comment (ebfull) [inline src/gtest/test_checktransaction.cpp]: See #862 for the ticket turning this into a mock-based test and removing the `SetPerformPourVerification` API in the future. (*Way* out of scope for this PR.)

**Comment by ebfull** [comment]
ACK

**Comment by defuse** [comment]
@zkbot r+

**Comment by ebfull** [comment]
Looks like this breaks sighash tests. You can fix it by only generating transactions that are have nonzero `vpub_old` **NAND** nonzero `vpub_new`. (Enable PRINT_JSON to print it out and replace the test vectors.)

**Comment by defuse** [comment]
For my future reference, to regenerate `src/test/data/sighash.json`:

1. Add `#define PRINT_SIGHASH_JSON` to `src/test/sighash_tests.cpp`
2. Rebuild.
3. Run `./src/test/test_bitcoin --run_test=sighash_tests/sighash_test | grep -v Running > src/test/data/sighash.json`
4. Remove the line added in (1) and rebuild.
5. Run `./src/test/test_bitcoin --run_test=sighash_tests` to make sure it works.

**Comment by ebfull** [comment]
Could you add cookie-cutter `test_merkletree.cpp` and `test_noteencryption.cpp` to the `gtest` test suite and I could rebase (#888 and #889) on top of that with linking fixes? As it stands my PRs will cause merge conflicts.

**Comment by defuse** [comment]
@sean: Like that?

**Comment by ebfull** [comment]
Well, I'm not sure that gtest will be happy if the test names are the same in the files.

**Comment by defuse** [comment]
Oh good point, let me fix that.

**Comment by ebfull** [comment]
ACK when that is fixed, and I think despite the fact this has already been reviewed let's get @daira or @str4d to check it out briefly if possible.

**Comment by defuse** [comment]
@sean: Done. (I rebased and force pushed)

**Comment by str4d** [comment]
LGTM

**Comment by ebfull** [comment]
@zkbot r+

**Comment by defuse** [comment]
Crap I should have just run the tests locally

**Comment by ebfull** [comment]
@zkbot r+

**Pull request merged**